### PR TITLE
Add rule allowing procedures to be injected, cache rule_for result

### PR DESCRIPTION
**Pull request by TheWudu**
Allow procedures within rules which can be injected from outside.

Why like this?
I want to add a rate limiting for bots (or whitelist) and i don't want to add a dependency of another gem like legitbot to rack-throttle. 

Basically the Proc can be anything based on the request.

### CODE REVIEW

File: lib/rack/throttle/rules.rb [hunk 1]
```diff
@@ -27,29 +27,33 @@ def default_limit
...
         rule = rule_for(request)
         rule ? rule[:whitelisted] : false
       end
 
       def rule_for(request)
-        rules.find do |rule|
-          next unless rule[:method] == request.request_method.to_s
-          next unless path_matches?(rule, request.path.to_s)
-          rule
+        cache.fetch("rule-#{request.object_id}") do
```
Comment (TheWudu) [inline lib/rack/throttle/rules.rb]: i added the caching as rule_for is called quite often and it is O(n) each time. There is no need to run through it that many times. Another option would be to rewrite the whole rack-throttle, to determine the cause once and pass it through the methods... as a cache is available, and a object-id should be unique somehow,.. maybe i should add an expire here after a few seconds..

File: rack-throttle.gemspec [hunk 2]
```diff
@@ -35,6 +35,8 @@ Gem::Specification.new do |gem|
   gem.add_development_dependency 'rspec'
   gem.add_development_dependency 'yard'
   gem.add_development_dependency 'timecop'
+  gem.add_development_dependency 'pry'
+  gem.add_development_dependency 'awesome_print'
```
Comment (nilsding) [inline rack-throttle.gemspec]: Are those gems _really_ required for development of this gem?

Comment (TheWudu) [reply rack-throttle.gemspec]: no they are not, forgot to remove it again .-)

**Review by nilsding — APPROVED** [review]

File: lib/rack/throttle/rules.rb [hunk 3]
```diff
@@ -46,14 +46,11 @@ def rule_whitelisted?(request)
       end
 
       def rule_for(request)
-        cache.fetch("rule-#{request.object_id}") do
```
Comment (nilsding) [inline lib/rack/throttle/rules.rb]: so... no more caching of the rules?

Comment (TheWudu) [reply lib/rack/throttle/rules.rb]: procs can't be marshaled into something cache-able, and only caching the index if the rule in the array seems somehow hacky.

**Pull request merged**